sql/uniques.h

```diff
@@ -16,20 +16,228 @@
 #ifndef UNIQUE_INCLUDED
 #define UNIQUE_INCLUDED
 
+#include "mariadb.h"
+#include "field.h"
 #include "filesort.h"
+#include "my_tree.h"
+#include "sql_string.h"
+
+
+/*
+
+  Keys_descriptor class storing information about the keys that would be
+  inserted in the Unique tree. This is an abstract class which is
+  extended by other class to support descriptors for keys with fixed and
+  variable size.
+*/
+
+class Keys_descriptor : public Sql_alloc
+{
+protected:
+
+  /* maximum possible size of any key, in bytes */
+  uint max_length;
+  enum attributes
+  {
+    FIXED_SIZED_KEYS= 0,
+    VARIABLE_SIZED_KEYS
+  } keys_type;
+
+  /*
+    Array of SORT_FIELD structure storing the information about the key parts
+    in the sort key of the Unique tree
+    @see Unique::setup()
+  */
+  SORT_FIELD *sortorder;
+
+  /*
+    Structure storing information about usage of keys
+  */
+  Sort_keys *sort_keys;
+
+public:
+  virtual ~Keys_descriptor() {};
+  virtual uint get_length_of_key(uchar *ptr) = 0;
+  bool is_variable_sized() { return keys_type == VARIABLE_SIZED_KEYS; }
+  virtual int compare_keys(const uchar *a, const uchar *b) const = 0;
+
+  // Fill structures like sort_keys, sortorder
+  virtual bool setup_for_item(THD *thd, Item_sum *item,
+                              uint non_const_args, uint arg_count);
+  virtual bool setup_for_field(THD *thd, Field *field);
+
+  uint get_max_key_length() const { return max_length; };
+  Sort_keys *get_keys() { return sort_keys; }
+  SORT_FIELD *get_sortorder() { return sortorder; }
+
+  /*
+    Create a keys record according to this descriptor's storage format.
+
+    The data is taken from the SORT_FIELD. The SORT_FIELD's values (underlying
+    table::record[0] must be set before calling this function.
+
+    The record is then to be used within a Unique object to store the result.
+  */
```

Comment:
"a Unique object"? Do you mean not this (`*this`) object but some other?

Reply:
This is not a Unique object class, it's a Keys_descriptor. It is indeed passed to the constructor of Unique, but the `Keys_descriptor` doesn't hold a reference to the Unique object, so no, not any `*this` object. This object is just a "binary-data-generator + comparison function for generated data"

Reply:
Ok, some leftover comments from Varun's patch make this confusing. I will reword them.